<?xml version="1.0"?>
<robot name="cart_pole" xmlns:xacro="http://www.ros.org/wiki/xacro">

  <!-- Constants -->
  <xacro:property name="cart_mass" value="1.0"/>
  <xacro:property name="pole_mass" value="0.1"/>
  <xacro:property name="pole_length" value="1.0"/>
  
  <!-- Base Link (World) -->
  <link name="world"/>
  
  <!-- Rail -->
  <link name="rail">
    <visual>
      <geometry>
        <box size="3.0 0.1 0.1"/>
      </geometry>
      <material name="gray">
        <color rgba="0.5 0.5 0.5 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <box size="3.0 0.1 0.1"/>
      </geometry>
    </collision>
  </link>

  <joint name="world_to_rail" type="fixed">
    <parent link="world"/>
    <child link="rail"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
  </joint>

  <!-- Cart -->
  <link name="cart">
    <inertial>
      <mass value="${cart_mass}"/>
      <inertia ixx="0.01" ixy="0.0" ixz="0.0" iyy="0.01" iyz="0.0" izz="0.01"/>
    </inertial>
    <visual>
      <geometry>
        <box size="0.3 0.2 0.1"/>
      </geometry>
      <material name="blue">
        <color rgba="0 0 1 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <box size="0.3 0.2 0.1"/>
      </geometry>
    </collision>
  </link>

  <joint name="cart_slider" type="prismatic">
    <parent link="rail"/>
    <child link="cart"/>
    <axis xyz="1 0 0"/>
    <origin xyz="0 0 0.1"/>
    <limit lower="-1.5" upper="1.5" effort="1000.0" velocity="1000.0"/>
    <dynamics damping="0.1" friction="0.1"/>
  </joint>

  <!-- Pole -->
  <link name="pole">
    <inertial>
      <mass value="${pole_mass}"/>
      <origin xyz="0 0 ${pole_length/2}"/>
      <inertia 
        ixx="${pole_mass * pole_length * pole_length / 12}" 
        ixy="0.0" 
        ixz="0.0" 
        iyy="${pole_mass * pole_length * pole_length / 12}" 
        iyz="0.0" 
        izz="0.001"/>
    </inertial>
    <visual>
      <origin xyz="0 0 ${pole_length/2}"/>
      <geometry>
        <cylinder radius="0.025" length="${pole_length}"/>
      </geometry>
      <material name="red">
        <color rgba="1 0 0 1"/>
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 ${pole_length/2}"/>
      <geometry>
        <cylinder radius="0.025" length="${pole_length}"/>
      </geometry>
    </collision>
  </link>

  <joint name="pole_hinge" type="revolute">
    <parent link="cart"/>
    <child link="pole"/>
    <axis xyz="0 1 0"/>
    <origin xyz="0 0 0.05"/>
    <limit lower="-3.14159" upper="3.14159" effort="1000.0" velocity="1000.0"/>
    <dynamics damping="0.1" friction="0.1"/>
  </joint>

  <!-- Gazebo plugins -->
  <gazebo>
    <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so">
      <robotNamespace>/cart_pole</robotNamespace>
      <robotSimType>gazebo_ros_control/DefaultRobotHWSim</robotSimType>
    </plugin>
  </gazebo>

  <!-- Gazebo colors -->
  <gazebo reference="rail">
    <material>Gazebo/Gray</material>
  </gazebo>
  <gazebo reference="cart">
    <material>Gazebo/Blue</material>
  </gazebo>
  <gazebo reference="pole">
    <material>Gazebo/Red</material>
  </gazebo>

  <!-- Transmissions for ROS control -->
  <transmission name="cart_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="cart_slider">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="cart_motor">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <transmission name="pole_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="pole_hinge">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="pole_motor">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

</robot> 